strinject.c

```diff
@@ -0,0 +1,462 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <limits.h>
+
+#include "common.h"
+#include "str.h"
+#include "malloc.h"
+#include "file.h"
+
+#include "strinject.h"
+
+static struct rtr_strinject_info g_strinject_infos[STRINJECT_FUNC_MAX];
+static int g_strinject_init;
+
+/*
+ * string inject types
+ */
+
+static struct {
+	enum RTR_STRINJECT_TYPE type;
+	const char *type_str;
+} g_inject_types[] = {
+	{STRINJECT_TYPE_HEX, "INJECT_SINGLE_HEX"},
+	{STRINJECT_TYPE_FMT_STR, "INJECT_FORMAT_STR"},
+	{STRINJECT_TYPE_BUF_OVERFLOW, "INJECT_BUF_OVERFLOW"},
+	{STRINJECT_TYPE_FILE_LINE, "INJECT_FILE_LINE"},
+	{STRINJECT_TYPE_UNKNOWN, NULL}
+};
+
+/*
+ * string inject functions
+ */
+
+static struct {
+	enum RTR_STRINJECT_FUNC_ID id;
+	const char *name;
+} g_inject_funcs[] = {
+	{STRINJECT_FUNC_FWRITE, "fwrite"},
+	{STRINJECT_FUNC_FREAD, "fread"},
+	{STRINJECT_FUNC_READ, "read"},
+	{STRINJECT_FUNC_WRITE, "write"},
+	{STRINJECT_FUNC_SEND, "send"},
+	{STRINJECT_FUNC_SENDTO, "sendto"},
+	{STRINJECT_FUNC_SENDMSG, "sendmsg"},
+	{STRINJECT_FUNC_RECV, "recv"},
+	{STRINJECT_FUNC_RECVFROM, "recvfrom"},
+	{STRINJECT_FUNC_RECVMSG, "recvmsg"},
+	{STRINJECT_FUNC_MAX, NULL}
+};
+
+/*
+ * string injection init function
+ */
+
+static void rtr_strinject_init(void)
+{
+	RTR_CONFIG_HANDLE config = RTR_CONFIG_START;
+
+	while (1) {
+		char *inject_type_str = NULL;
+		char *func_list = NULL;
+		char *inject_param = NULL;
+		double inject_rate;
+
+		enum RTR_STRINJECT_TYPE inject_type = STRINJECT_TYPE_UNKNOWN;
+
+		int i, reverse;
+
+		/* get configuration line */
+		if (rtr_get_config_multiple(&config, "stringinject", ARGUMENT_TYPE_STRING, ARGUMENT_TYPE_STRING,
+			ARGUMENT_TYPE_STRING, ARGUMENT_TYPE_DOUBLE, ARGUMENT_TYPE_END,
+			&inject_type_str, &func_list, &inject_param, &inject_rate) == 0)
+			break;
+
+		/* get inject type */
+		for (i = 0; i < STRINJECT_TYPE_UNKNOWN; i++) {
+			if (real_strcmp(inject_type_str, g_inject_types[i].type_str) == 0) {
+				inject_type = g_inject_types[i].type;
+				break;
+			}
+		}
+
+		if (inject_type == STRINJECT_TYPE_UNKNOWN)
+			continue;
+
+		/* get function list */
+		for (i = 0; i < STRINJECT_FUNC_MAX; i++) {
+			if (rtr_check_config_token(g_inject_funcs[i].name, func_list, "|", &reverse) &&
+				!g_strinject_infos[i].exist) {
+				g_strinject_infos[i].type = inject_type;
+				real_strcpy(g_strinject_infos[i].param, inject_param);
+				g_strinject_infos[i].rate = inject_rate;
+
+				g_strinject_infos[i].exist = 1;
+			}
+		}
+	}
+
+	/* set init flag */
+	g_strinject_init = 1;
+}
+
+/*
+ * parse injection parameter
+ */
+
+static int parse_inject_param(enum RTR_STRINJECT_TYPE type, const char *param, size_t len, void *val, off_t *offset)
+{
+	char param_val[512];
+	char *p;
+
+	int fmt_count, overflow_buf_len;
+
+	/* parse param */
+	p = real_strchr(param, ':');
+	if (!p)
+		return -1;
+
+	memset(param_val, 0, sizeof(param_val));
+	real_strncpy(param_val, param, p - param);
```

Comment:
To avoid possible overflow:
```
if (!p || p - param >= 512)
    return -1;

real_strncpy(param_val, param, p - param);
param[p - param] = '\0';
```